adios2 current master is broken for  -DADIOS2_USE_Endian_Reverse=ON

Related to changes in types from @germasch . See below for compilation with ninja (make has similar results).
@chuckatkins should we add tests with big-endian files (I can provide) to make sure ADIOS2_USE_Endian_Reverse=ON doesn't break?

```
ninja
Recompacting log...
[6/212] Linking CXX executable bin/adios_reorganize
FAILED: : && /usr/bin/c++  -g   source/utils/CMakeFiles/adios_reorganize.dir/adios_reorganize/main.cpp.o source/utils/CMakeFiles/adios_reorganize.dir/adios_reorganize/Reorganize.cpp.o source/utils/CMakeFiles/adios_reorganize.dir/Utils.cpp.o  -o bin/adios_reorganize  -Wl,-rpath,/home/wgodoy/workspace/build/lib:/usr/lib/openmpi/lib: lib/libadios2.so.2.3.1 -Wl,-rpath -Wl,/usr/lib/openmpi/lib -Wl,--enable-new-dtags -pthread /usr/lib/openmpi/lib/libmpi.so -lpthread && :
lib/libadios2.so.2.3.1: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > adios2::helper::GetType<long long>()'
lib/libadios2.so.2.3.1: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > adios2::helper::GetType<char>()'
lib/libadios2.so.2.3.1: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > adios2::helper::GetType<unsigned long long>()'
collect2: error: ld returned 1 exit status
[6/212] Linking CXX executable bin/adios_iotest
FAILED: : && /usr/bin/c++  -g   source/utils/adios_iotest/CMakeFiles/adios_iotest.dir/settings.cpp.o source/utils/adios_iotest/CMakeFiles/adios_iotest.dir/decomp.cpp.o source/utils/adios_iotest/CMakeFiles/adios_iotest.dir/processConfig.cpp.o source/utils/adios_iotest/CMakeFiles/adios_iotest.dir/ioGroup.cpp.o source/utils/adios_iotest/CMakeFiles/adios_iotest.dir/stream.cpp.o source/utils/adios_iotest/CMakeFiles/adios_iotest.dir/adiosStream.cpp.o source/utils/adios_iotest/CMakeFiles/adios_iotest.dir/adios_iotest.cpp.o source/utils/adios_iotest/CMakeFiles/adios_iotest.dir/hdf5Stream.cpp.o  -o bin/adios_iotest  -Wl,-rpath,/home/wgodoy/workspace/build/lib:/usr/lib/x86_64-linux-gnu/hdf5/openmpi:/usr/lib/openmpi/lib: lib/libadios2.so.2.3.1 /usr/lib/x86_64-linux-gnu/hdf5/openmpi/libhdf5.so /usr/lib/x86_64-linux-gnu/libsz.so /usr/lib/x86_64-linux-gnu/libz.so /usr/lib/x86_64-linux-gnu/libdl.so -lm -Wl,-rpath -Wl,/usr/lib/openmpi/lib -Wl,--enable-new-dtags -pthread /usr/lib/openmpi/lib/libmpi.so -lpthread && :
lib/libadios2.so.2.3.1: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > adios2::helper::GetType<long long>()'
lib/libadios2.so.2.3.1: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > adios2::helper::GetType<char>()'
lib/libadios2.so.2.3.1: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > adios2::helper::GetType<unsigned long long>()'
collect2: error: ld returned 1 exit status
[6/212] Linking CXX executable bin/bpls
FAILED: : && /usr/bin/c++  -g   thirdparty/KWSys/adios2sys/CMakeFiles/adios2sys_objects.dir/EncodingC.c.o thirdparty/KWSys/adios2sys/CMakeFiles/adios2sys_objects.dir/Directory.cxx.o thirdparty/KWSys/adios2sys/CMakeFiles/adios2sys_objects.dir/DynamicLoader.cxx.o thirdparty/KWSys/adios2sys/CMakeFiles/adios2sys_objects.dir/EncodingCXX.cxx.o thirdparty/KWSys/adios2sys/CMakeFiles/adios2sys_objects.dir/RegularExpression.cxx.o thirdparty/KWSys/adios2sys/CMakeFiles/adios2sys_objects.dir/SystemTools.cxx.o thirdparty/KWSys/adios2sys/CMakeFiles/adios2sys_objects.dir/CommandLineArguments.cxx.o thirdparty/KWSys/adios2sys/CMakeFiles/adios2sys_objects.dir/FStream.cxx.o source/utils/CMakeFiles/bpls.dir/bpls/bpls.cpp.o source/utils/CMakeFiles/bpls.dir/__/__/thirdparty/pugixml/pugixml/src/pugixml.cpp.o  -o bin/bpls  -Wl,-rpath,/home/wgodoy/workspace/build/lib:/usr/lib/openmpi/lib: lib/libadios2.so.2.3.1 -lpthread -Wl,-rpath -Wl,/usr/lib/openmpi/lib -Wl,--enable-new-dtags -pthread /usr/lib/openmpi/lib/libmpi.so -ldl && :
lib/libadios2.so.2.3.1: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > adios2::helper::GetType<long long>()'
lib/libadios2.so.2.3.1: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > adios2::helper::GetType<char>()'
lib/libadios2.so.2.3.1: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > adios2::helper::GetType<unsigned long long>()'
collect2: error: ld returned 1 exit status
[6/212] Building CXX object bindings/Python/CMakeFiles/adios2py.dir/py11File.cpp.o
ninja: build stopped: subcommand failed.
```

I can look at this tonight, it looks familiar

On Mon, Mar 4, 2019, 1:34 PM William F Godoy <notifications@github.com>
wrote:

> Related to changes in types. See below for compilation with ninja (make
> has similar results).
> @chuckatkins <https://github.com/chuckatkins> should we add tests with
> big-endian files (I can provide) to make sure ADIOS2_USE_Endian_Reverse=ON
> doesn't break?
>
> ninja
> Recompacting log...
> [6/212] Linking CXX executable bin/adios_reorganize
> FAILED: : && /usr/bin/c++  -g   source/utils/CMakeFiles/adios_reorganize.dir/adios_reorganize/main.cpp.o source/utils/CMakeFiles/adios_reorganize.dir/adios_reorganize/Reorganize.cpp.o source/utils/CMakeFiles/adios_reorganize.dir/Utils.cpp.o  -o bin/adios_reorganize  -Wl,-rpath,/home/wgodoy/workspace/build/lib:/usr/lib/openmpi/lib: lib/libadios2.so.2.3.1 -Wl,-rpath -Wl,/usr/lib/openmpi/lib -Wl,--enable-new-dtags -pthread /usr/lib/openmpi/lib/libmpi.so -lpthread && :
> lib/libadios2.so.2.3.1: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > adios2::helper::GetType<long long>()'
> lib/libadios2.so.2.3.1: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > adios2::helper::GetType<char>()'
> lib/libadios2.so.2.3.1: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > adios2::helper::GetType<unsigned long long>()'
> collect2: error: ld returned 1 exit status
> [6/212] Linking CXX executable bin/adios_iotest
> FAILED: : && /usr/bin/c++  -g   source/utils/adios_iotest/CMakeFiles/adios_iotest.dir/settings.cpp.o source/utils/adios_iotest/CMakeFiles/adios_iotest.dir/decomp.cpp.o source/utils/adios_iotest/CMakeFiles/adios_iotest.dir/processConfig.cpp.o source/utils/adios_iotest/CMakeFiles/adios_iotest.dir/ioGroup.cpp.o source/utils/adios_iotest/CMakeFiles/adios_iotest.dir/stream.cpp.o source/utils/adios_iotest/CMakeFiles/adios_iotest.dir/adiosStream.cpp.o source/utils/adios_iotest/CMakeFiles/adios_iotest.dir/adios_iotest.cpp.o source/utils/adios_iotest/CMakeFiles/adios_iotest.dir/hdf5Stream.cpp.o  -o bin/adios_iotest  -Wl,-rpath,/home/wgodoy/workspace/build/lib:/usr/lib/x86_64-linux-gnu/hdf5/openmpi:/usr/lib/openmpi/lib: lib/libadios2.so.2.3.1 /usr/lib/x86_64-linux-gnu/hdf5/openmpi/libhdf5.so /usr/lib/x86_64-linux-gnu/libsz.so /usr/lib/x86_64-linux-gnu/libz.so /usr/lib/x86_64-linux-gnu/libdl.so -lm -Wl,-rpath -Wl,/usr/lib/openmpi/lib -Wl,--enable-new-dtags -pthread /usr/lib/openmpi/lib/libmpi.so -lpthread && :
> lib/libadios2.so.2.3.1: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > adios2::helper::GetType<long long>()'
> lib/libadios2.so.2.3.1: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > adios2::helper::GetType<char>()'
> lib/libadios2.so.2.3.1: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > adios2::helper::GetType<unsigned long long>()'
> collect2: error: ld returned 1 exit status
> [6/212] Linking CXX executable bin/bpls
> FAILED: : && /usr/bin/c++  -g   thirdparty/KWSys/adios2sys/CMakeFiles/adios2sys_objects.dir/EncodingC.c.o thirdparty/KWSys/adios2sys/CMakeFiles/adios2sys_objects.dir/Directory.cxx.o thirdparty/KWSys/adios2sys/CMakeFiles/adios2sys_objects.dir/DynamicLoader.cxx.o thirdparty/KWSys/adios2sys/CMakeFiles/adios2sys_objects.dir/EncodingCXX.cxx.o thirdparty/KWSys/adios2sys/CMakeFiles/adios2sys_objects.dir/RegularExpression.cxx.o thirdparty/KWSys/adios2sys/CMakeFiles/adios2sys_objects.dir/SystemTools.cxx.o thirdparty/KWSys/adios2sys/CMakeFiles/adios2sys_objects.dir/CommandLineArguments.cxx.o thirdparty/KWSys/adios2sys/CMakeFiles/adios2sys_objects.dir/FStream.cxx.o source/utils/CMakeFiles/bpls.dir/bpls/bpls.cpp.o source/utils/CMakeFiles/bpls.dir/__/__/thirdparty/pugixml/pugixml/src/pugixml.cpp.o  -o bin/bpls  -Wl,-rpath,/home/wgodoy/workspace/build/lib:/usr/lib/openmpi/lib: lib/libadios2.so.2.3.1 -lpthread -Wl,-rpath -Wl,/usr/lib/openmpi/lib -Wl,--enable-new-dtags -pthread /usr/lib/openmpi/lib/libmpi.so -ldl && :
> lib/libadios2.so.2.3.1: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > adios2::helper::GetType<long long>()'
> lib/libadios2.so.2.3.1: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > adios2::helper::GetType<char>()'
> lib/libadios2.so.2.3.1: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > adios2::helper::GetType<unsigned long long>()'
> collect2: error: ld returned 1 exit status
> [6/212] Building CXX object bindings/Python/CMakeFiles/adios2py.dir/py11File.cpp.o
> ninja: build stopped: subcommand failed.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ornladios/ADIOS2/issues/1253>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AKgzRqZ_J6dY2qIAVkU6Q-p38_4Q15qbks5vTWcXgaJpZM4bc87H>
> .
>

--- a/source/adios2/helper/adiosMemory.cpp
+++ b/source/adios2/helper/adiosMemory.cpp
@@ -38,7 +38,7 @@ void CopyPayloadStride(const char *src, const size_t payloadStride, char *dest,
         CopyEndianReverse<T>(src, payloadStride, reinterpret_cast<T *>(dest)); \
     }
 
-        ADIOS2_FOREACH_PRIMITIVE_TYPE_1ARG(declare_type)
+        ADIOS2_FOREACH_PRIMITIVE_STDTYPE_1ARG(declare_type)
 #undef declare_type
     }
     else
@@ -285,6 +285,11 @@ void CopyPayload(char *dest, const Dims &destStart, const Dims &destCount,
                         srcMemCount, endianReverse, destType);
     }
 }
+
+#ifdef ADIOS2_HAVE_ENDIAN_REVERSE
+template void CopyEndianReverse(const char *src, const size_t payloadStride,
+                                char *dest);
+#endif
 
 } // end namespace helper
 } // end namespace adios2

